dot/sync/utils.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package sync
+
+type reducer func(prevValue interface{}, newValue interface{}) interface{}
+
+// for comp: comp(x, y) = -1: x<y, 0: x==y, 1: x>y
+type comperator func(prevValue interface{}, newValue interface{}) int
+
+//RemoveOutlier removes the outlier from the slice
+//  Explanation:
+// 	IQR outlier detection
+// 	Q25 = 25th_percentile
+// 	Q75 = 75th_percentile
+// 	IQR = Q75 - Q25         // inter-quartile range
+// 	If x >  Q75  + 1.5 * IQR or  x   < Q25 - 1.5 * IQR THEN  x is a mild outlier
+// 	If x >  Q75  + 3.0 * IQR or  x   < Q25 – 3.0 * IQR THEN  x is a extreme outlier
+// Ref: http://www.mathwords.com/o/outlier.htm
+// returns: reducer output
+func RemoveOutlier(sortedArr []interface{}, compFn comperator, initialReducedVal interface{}, reducer, plusFn, minusFn, divideFn, multiplyFn reducer) interface{} {
```

Comment:
does this need to be exported?

Reply:
also, if we're only passing `*big.Int`s into this function, can we just make it take `*big.Int` and return `*big.Int` instead of `intertface{}`? this seems unnecessarily complex with the interfaces :p

Reply:
also I don't entirely understand what this is returning... is it the sum of all the values, with the outliers removed?

Reply:
@noot I was thinking of making the Outlier remover method generic so that we could use it other slice type than big.Int in future. Should I yet make big.Int? And exported the method to be useable from other package if needed.

Reply:
@noot `is it the sum of all the values, with the outliers removed?`: Yes

Reply:
> @noot I was thinking of making the Outlier remover method generic so that we could use it in other slice types than big.Int in future. Should I yet make big.Int? And exporting the method to be useable from other packages if needed.

I don't think that's necessary since it's a pre-optimization we might not ever need, I think it should only be `*big.Int` and doesn't need to be exported

> @noot is it the sum of all the values, with the outliers removed?: Yes
 
ok cool, can you update the comment to say this?

Reply:
done ✔